Source/Dafny/Compiler.cs

```diff
@@ -627,6 +630,8 @@ public enum ResolvedUnaryOp { BoolNot, BitwiseNot, Cardinality }
     protected abstract void EmitSingleValueGenerator(Expression e, bool inLetExprBody, string type, TargetWriter wr);
     protected virtual void FinishModule() { }
 
+    protected virtual void DeclareExternType(OpaqueTypeDecl d, Expression compileTypeHint, TargetWriter wr) { }
```

Comment:
What is the exact definition of this method? Externs are in a fairly fuzzy state right now (eg #469) so I'd like to understand the use cases for this in C++ in particular and how they generalize to other target languages.

Reply:
Here's what the C++ backend uses it for:

`    protected override void DeclareExternType(OpaqueTypeDecl d, Expression compileTypeHint, TargetWriter wr) {
      if (compileTypeHint.AsStringLiteral() == "struct") {
        modDeclWr.WriteLine("// Extern declaration of {1}\n{0} struct {1} {2};", DeclareTemplate(d.TypeArgs), d.Name, TemplateMethod(d.TypeArgs));        
      } else if (compileTypeHint.AsStringLiteral() == "predefined") {
        // This type already exists, so no declaration needed
      } else {
        Error(d.tok, "Opaque type ('{0}') with unrecognized extern attribute {1} cannot be compiled.  Expected {{:extern compile_type_hint}} ", wr, d.FullName, compileTypeHint.AsStringLiteral());
      }
    }`

It's trying to handle two specific issues that came up in our C++ code, both influenced by our current handling of include files.  At the moment, the user can pass one or more `.h` files to Dafny when compiling for C++, and Dafny will add them to the generated `.cpp` file, but not the `.h` file, in the hopes of keeping the generated files relatively self-contained.  I'm not wedded to this approach, but that's what we have at the moment.  The `struct` case above was necessary for a project that wanted to include the generated `.h` in another file and needed a basic definition for the extern type.  The `predefined` case, on the other hand, is for when we already have a complete (possibly non-struct) drop-in type we want to use.

I agree about extern being in a rather awkward state, so I'm all for finding a cleaner, more coherent model for them.

Reply:
Got it. I feel pretty strongly that Dafny should have two separate mechanisms for declaring when Dafny code should create externally-visible new elements as opposed to importing existing ones. I'm planing on proposing that fairly soon and I'll keep this use case in mind. Hopefully that means your compile hint can be a more generic attribute instead.

Reply:
Yes, I agree that splitting those two semantically different roles out would help.

Reply:
You're using an attribute like `{:extern "predefined"}` to indicate a type already exists, correct?

My only concern is that this is ambiguous with the existing usage of this attribute, where the arguments are used to  provide the symbol(s) used in the compiled version (e.g. . I don't think we want the arguments to have different meanings for different compilers, so we should resolve this somehow.

I'm focusing on a proper solution to #469 this week. I have a deeper solution to configuring the mappings/bindings for each language target in mind, but in the short-term I'd like to just have two different attributes: one for "make the compiled version of this thing accessible in the target language" vs. "the compiled version already exists, just declare this symbol as a Dafny alias for it".

Would that meet your needs here, or do you anticipate having more values for this hint? Other configurations and optimizations will hopefully fit into my more long term proposal.

Reply:
Yes, that's correct.  "Predefined" means it already exists so don't even declare it.  "struct" says give a very basic "struct type_name" definition.  But you make a good point that this isn't consistent with the other uses of arguments to extern.  I think the two use cases you outline above for the short term would likely suffice for our purposes.

Reply:
If I'm reading between the lines correctly :) you have at least one other option in mind for how to compile a Dafny type to C++? I'm hoping that the solution I propose for #469 and some related issues would also cover that kind of compiler customization - it's a slippery slope from just specifying the names you want in each language to tweaking a lot more than that. If you don't mind describing potential future "hints" in more detail it would help me out a lot!

Reply:
At the moment, the C++ backend compiles Dafny datatypes to C++ structs, and Dafny classes to C++ classes.  Thus far, the code we're interfacing with at the native level has mostly been C code that hands us (and expects) structs, so we've modeled them in Dafny as datatype-like types (and used the `:extern "struct"` to declare them properly).  In the future, I could imagine interfacing with C++ objects, which we'd likely want to model via Dafny classes and export as such.